improver/field_texture.py

```diff
@@ -0,0 +1,251 @@
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+""" Module containing plugin to calculate whether or not the input field texture
+    exceeds a given threshold."""
+
+import iris
+import numpy as np
+from iris.exceptions import CoordinateNotFoundError
+
+from improver import BasePlugin
+from improver.metadata.probabilistic import find_threshold_coordinate
+from improver.metadata.utilities import (
+    create_new_diagnostic_cube,
+    generate_mandatory_attributes,
+)
+from improver.nbhood.square_kernel import SquareNeighbourhood
+from improver.threshold import BasicThreshold
+from improver.utilities.cube_manipulation import collapsed
+
+
+class FieldTexture(BasePlugin):
+    """Plugin to calculate whether or not the input field texture exceeds a
+       given threshold.
+
+    1) Takes a binary field that has been thresholded and looks for the transitions/edges
+       in the field that mark out a transition.
+    2) The transition calculation is then fed into the neighbourhooding code
+       (_calculate_ratio) to calculate a ratio for each cell. This is the texture
+       of the input field.
+    3) The new cube of ratios is then thresholded and the realization coordinates
+       are collapsed to generate a mean of the thresholded ratios. This gives a binary
+       indication of whether a field is rough (texture values close to 1) or smooth
+       (texture values close to zero).
+    """
+
+    def __init__(self, nbhood_radius, textural_threshold, diagnostic_threshold):
+        """
+
+        Args:
+            nbhood_radius (float):
+                The neighbourhood radius in metres within which the number of potential
+                transitions should be calculated. This forms the denominator in the
+                calculation of the ratio of actual to potential transitions that indicates a
+                field's texture. A larger radius should be used for diagnosing larger-scale
+                textural features.
+
+            textural_threshold (float):
+                A unit-less threshold value that defines the ratio value above which
+                the field is considered rough and below which the field is considered
+                smoother.
+
+            diagnostic_threshold (float):
+                A user defined threshold value related either to cloud or precipitation,
+                used to extract the corresponding dimensional cube (default: 1).
```

Comment:
This doesn't have a default value at the moment.  The assumed units are 1, and we should state this.

Reply:
Done.